### PR TITLE
OTel Agent Feature Support

### PR DESCRIPTION
**Pull request by mackjmr**
### What does this PR do?

- Add support for Otel Agent as a Feature
- Remove support for OTel Agent as a flag
- Maintains support for OTel Agent as an annotation _temporarily_

### Motivation

OTEL-2290

### Additional Notes

Anything else we should know when reviewing?

### Minimum Agent Versions

Are there minimum versions of the Datadog Agent and/or Cluster Agent required?

* Agent: vX.Y.Z
* Cluster Agent: vX.Y.Z

### Describe your test plan

This PR adds support for deploying the otel-agent in a new manner. In order to confirm these changes work as expected, we need to validate that the otel agent runs without error, but also that it is able to receive data from OTel clients and forward it to Datadog. 

The following [docker image](https://hub.docker.com/layers/mackjmr/app-all/v2/images/sha256-e36218f126213b5523fd1ddf1dec4234cd66b39a50ff086852ad88f4c282c36e): `mackjmr/app-all:v2` contains an app that sends OTLP traces every 5 seconds, which we can use to send data to the otel agent. 

This app can be deployed using the following `deployment.yaml`:
```
apiVersion: apps/v1
kind: Deployment
metadata:
  name: app-all
  labels:
    app: app-all
spec:
  replicas: 1
  selector:
    matchLabels:
      app: app-all
  template:
    metadata:
      labels:
        app: app-all
    spec:
      containers:
      - name: app-all
        image: mackjmr/app-all:v2
        env:
        - name: MY_NODE_IP
          valueFrom:
            fieldRef:
              fieldPath: status.hostIP
        - name: OTEL_EXPORTER_OTLP_ENDPOINT
          value: http://$(MY_NODE_IP):4317
        - name: POD_IP
          valueFrom:
            fieldRef:
              fieldPath: status.podIP
        - name: OTEL_RESOURCE_ATTRIBUTES
          value: "k8s.pod.ip=$(POD_IP),service.name=test-operator-changes"
```

*Noteworthy:*
- `OTEL_EXPORTER_OTLP_ENDPOINT` is sent to send traces to the NODE_IP:4317 (otel agent listens on 4317 and binds container port to host port)
- The service name is set via env var `OTEL_RESOURCE_ATTRIBUTES`.

*Known issues when testing:*
Agent will fail to startup in some kubernetes environments (e.g. Kind cluster) if it fails to detect the hostname. You can prevent this from happening by setting the hostname yourself:
```
spec:
  override:
    nodeAgent:
      containers: 
        agent:
          env:
            - name: DD_HOSTNAME
              value: "test.node.name"
```

In Kind/ Minikube, you also need to disable TLS Verify:
```
  global:
    kubelet:
      tlsVerify: false
```

### 1. Otel Agent Feature with collector config in configData:

*Test 1: Default ports*

Deploy the agent using the following manifest: https://github.com/DataDog/datadog-operator/blob/1a69194e7c958641288a3f0489372acb2cecfa34/examples/datadogagent/datadog-agent-with-otel-agent.yaml. 

- Ensure that the otel-agent starts up without issues/ errors. 
- Once above has been validated, deploy the app-all image mentionned above, and ensure that traces are making it into Datadog

*Test 2: Non default ports:*

Same steps as above, but in manifest add ports sections:
```
  features:
    otelCollector:
      enabled: true
      ports:
        - containerPort: 3333
          name: otel-grpc
``` 

and change collector config port:
```
    receivers:
      otlp:
        protocols:
          grpc:
            endpoint: 0.0.0.0:3333
```

and in app-all deployment change `OTEL_EXPORTER_OTLP_ENDPOINT` to  `http://$(MY_NODE_IP):3333`.

### 2. Otel Agent Feature with collector config in configMap:

Deploy the agent using the following manifest: https://github.com/DataDog/datadog-operator/blob/1a69194e7c958641288a3f0489372acb2cecfa34/examples/datadogagent/datadog-agent-with-otel-agent-configmap.yaml. 

- Ensure that the otel-agent starts up without issues/ errors. 
- Once above has been validated, deploy the app-all image mentionned above, and ensure that traces are making it into Datadog

### 3. Otel Agent Feature without passing collector config:

When we don't pass a collector config, a default config is used.

*Test 1: Default ports*

Deploy the agent using the following:
```
apiVersion: datadoghq.com/v2alpha1
kind: DatadogAgent
metadata:
  name: datadog
spec:
  global:
    credentials:
      apiKey: <DATADOG_API_KEY>
  features:
    otelCollector:
      enabled: true
```

- Ensure that the otel-agent starts up without issues/ errors. 
- Once above has been validated, deploy the app-all image mentionned above, and ensure that traces are making it into Datadog

*Test 2: Non default ports:*

Deploy the agent using the following:
```
apiVersion: datadoghq.com/v2alpha1
kind: DatadogAgent
metadata:
  name: datadog
spec:
  global:
    credentials:
      apiKey: <DATADOG_API_KEY>
  features:
    otelCollector:
      enabled: true
      ports:
        - containerPort: 3333
          name: otel-grpc
```

- Ensure that the otel-agent starts up without issues/ errors. 
- Once above has been validated, deploy the app-all image mentionned above *changing OTEL_EXPORTER_OTLP_ENDPOINT to  http://$(MY_NODE_IP):3333*, and ensure that traces are making it into Datadog

### 4. Otel Agent via Annotation:

We need to validate that deploying otel agent with annotations still works. 

Deploy the agent using the following manifest: https://github.com/DataDog/datadog-operator/blob/1a69194e7c958641288a3f0489372acb2cecfa34/examples/datadogagent/datadog-agent-with-otel-agent-annotations.yaml. 

- Ensure that the otel-agent starts up without issues/ errors. 
- Once above has been validated, deploy the app-all image mentionned above, and ensure that traces are making it into Datadog

### 5. Flare:

*Test 1: Core config disabled*

Disable the core config:
```
  features:
    otelCollector:
      coreConfig: 
        enabled: false
```

Trigger a flare (you can check the flare locally, or if you want to submit it, you can use ticket #1970459), and ensure that in `otel/otel-agent.log` you see log:
```
'otelcollector.enabled' is disabled in the configuration
```

*Test 2: Core config enabled*

The `otelcollector.enabled` should be enabled by default when the otel collector feature is enabled, so you can leave the coreConfig empty (this config: https://github.com/DataDog/datadog-operator/blob/1a69194e7c958641288a3f0489372acb2cecfa34/examples/datadogagent/datadog-agent-with-otel-agent.yaml will do). Trigger a flare (you can check the flare locally, or if you want to submit it, you can use ticket #1970459), and ensure that the otel directory contains the following data:

*Test 3: Core config other params*

Add the following params:
```
  features:
    otelCollector:
      coreConfig: 
        enabled: true
        extension_timeout: 13
        extension_url: "https://localhost:7777"
```

describe the pod (`kubectl describe pod -n <namespace> <pod_name>`), and ensure the following env vars are set:
```
      DD_OTELCOLLECTOR_ENABLED:                              true
      DD_OTELCOLLECTOR_EXTENSION_TIMEOUT:                    13
      DD_OTELCOLLECTOR_EXTENSION_URL:                        https://localhost:7777
```

### Checklist

- [ ] PR has at least one valid label: `bug`, `enhancement`, `refactoring`, `documentation`, `tooling`, and/or `dependencies`
- [ ] PR has a milestone or the `qa/skip-qa` label

### CODE REVIEW

File: internal/controller/datadogagent/component/agent/default.go [hunk 1]
```diff
@@ -397,23 +407,9 @@ func envVarsForSecurityAgent(dda metav1.Object) []corev1.EnvVar {
...
-	envs := []corev1.EnvVar{
-		// TODO: add additional env vars here
-	}
-
-	return append(envs, commonEnvVars(dda)...)
-}
-
 func volumeMountsForInitConfig() []corev1.VolumeMount {
 	return []corev1.VolumeMount{
-		common.GetVolumeMountForLogs(),
```
Comment (levan-m) [inline internal/controller/datadogagent/component/agent/default.go:410]: Why delete initContainer configs, is this even in otel agent code path?

File: api/datadoghq/v2alpha1/datadogagent_types.go [hunk 2]
```diff
@@ -44,6 +44,8 @@ type DatadogAgentSpec struct {
 type DatadogFeatures struct {
 	// Application-level features
 
+	// OTelAgent configuration.
+	OTelAgent *OTelAgentFeatureConfig `json:"otelAgent,omitempty"`
```
Comment (levan-m) [inline api/datadoghq/v2alpha1/datadogagent_types.go:48]: is `otelAgent` really best name for this feature? should this be `otelCollector` or just `otel`? other features don't mention 'Agent'.

Comment (mackjmr) [reply api/datadoghq/v2alpha1/datadogagent_types.go]: Renamed feature from `otelAgent` to `otelCollector`.

File: internal/controller/datadogagent/component/agent/default.go (hunk 1 above)
Comment (mackjmr) [reply internal/controller/datadogagent/component/agent/default.go:410]: This was a mistake, meant to remove for otel agent voluments. undid the change

**Review by aliciascott — APPROVED** [review]

File: api/datadoghq/v2alpha1/test/builder.go [hunk 3]
```diff
@@ -374,6 +375,50 @@ func (builder *DatadogAgentBuilder) WithProcessDiscoveryEnabled(enabled bool) *D
...
+}
+
+func (builder *DatadogAgentBuilder) WithOTelCollectorConfigMap() *DatadogAgentBuilder {
+	builder.datadogAgent.Spec.Features.OtelCollector.Conf = &v2alpha1.CustomConfig{}
+	builder.datadogAgent.Spec.Features.OtelCollector.Conf.ConfigMap = &v2alpha1.ConfigMapConfig{
+		Name: "user-provided-config-map",
+	}
+	return builder
+}
+
```
Comment (dineshg13) [inline api/datadoghq/v2alpha1/test/builder.go:405]: Shouldn't the operator create the config map for user provided config ? Are we asking to provide configmap as input ?

File: cmd/main.go [hunk 4]
```diff
@@ -157,7 +156,6 @@ func (opts *options) Parse() {
 	flag.BoolVar(&opts.introspectionEnabled, "introspectionEnabled", false, "Enable introspection (beta)")
 	flag.BoolVar(&opts.datadogAgentProfileEnabled, "datadogAgentProfileEnabled", false, "Enable DatadogAgentProfile controller (beta)")
 	flag.BoolVar(&opts.remoteConfigEnabled, "remoteConfigEnabled", false, "Enable RemoteConfig capabilities in the Operator (beta)")
-	flag.BoolVar(&opts.otelAgentEnabled, "otelAgentEnabled", false, "Enable the OTel agent container (beta)")
```
Comment (dineshg13) [inline cmd/main.go:160]: Isn't this breaking change for staging?

File: config/crd/bases/v1/datadoghq.com_datadogagents.yaml [hunk 5]
```diff
@@ -1284,6 +1284,108 @@ spec:
                             Default: true
                           type: boolean
                       type: object
+                    otelCollector:
+                      description: OtelCollector configuration.
+                      properties:
+                        conf:
```
Comment (dineshg13) [inline config/crd/bases/v1/datadoghq.com_datadogagents.yaml:1290]: rename to config?

Comment (mackjmr) [reply config/crd/bases/v1/datadoghq.com_datadogagents.yaml:1290]: Same as: https://github.com/DataDog/datadog-operator/pull/1559#discussion_r1886798434

File: api/datadoghq/v2alpha1/test/builder.go (hunk 3 above)
Comment (mackjmr) [reply api/datadoghq/v2alpha1/test/builder.go:405]: We support two ways to pass config:
- Pass configmap directly: https://github.com/DataDog/datadog-operator/blob/1a69194e7c958641288a3f0489372acb2cecfa34/api/datadoghq/v2alpha1/test/builder.go#L398 => we don't need to build configmap
- Pass config via configdata: https://github.com/DataDog/datadog-operator/blob/1a69194e7c958641288a3f0489372acb2cecfa34/api/datadoghq/v2alpha1/test/builder.go#L391 => in this path we create the config map for the user

File: cmd/main.go (hunk 4 above)
Comment (mackjmr) [reply cmd/main.go:160]: My understanding is that staging uses annotations, not the flag so this shouldn't be a breaking change (I left annotation support for now, until we update staging to use feature, in which case I will remove annotation support).

cc @levan-m to confirm.

File: api/datadoghq/v2alpha1/test/builder.go (hunk 3 above)
Comment (mackjmr) [reply api/datadoghq/v2alpha1/test/builder.go:405]: Example passing configmap: https://github.com/DataDog/datadog-operator/blob/1a69194e7c958641288a3f0489372acb2cecfa34/examples/datadogagent/datadog-agent-with-otel-agent-configmap.yaml

Example passing configData (we create configMap): https://github.com/DataDog/datadog-operator/blob/1a69194e7c958641288a3f0489372acb2cecfa34/examples/datadogagent/datadog-agent-with-otel-agent.yaml

File: internal/controller/datadogagent/feature/otelcollector/feature.go [hunk 6]
```diff
@@ -0,0 +1,169 @@
...
+			},
+			{
+				Name:          "otel-grpc",
+				ContainerPort: 4317,
+				HostPort:      4317,
+				Protocol:      corev1.ProtocolTCP,
+			},
+		}
+	}
+	o.ports = dda.Spec.Features.OtelCollector.Ports
```
Comment (levan-m) [inline internal/controller/datadogagent/feature/otelcollector/feature.go:69]: Why not assign the configured ports' structure directly to `o.ports`? As a general practice, we don't modify DDA withing the features, instead save relevant config in feature struct `otelCollectorFeature`.

Comment (mackjmr) [reply internal/controller/datadogagent/feature/otelcollector/feature.go]: Makes sense, updated in: https://github.com/DataDog/datadog-operator/pull/1559/commits/ff7868ebd53219d2ebb593197edf50c5f2584bbf.

File: internal/controller/datadogagent/component/agent/default.go [hunk 7]
```diff
@@ -104,6 +104,11 @@ func agentImage() string {
 	return fmt.Sprintf("%s/%s:%s", v2alpha1.DefaultImageRegistry, v2alpha1.DefaultAgentImageName, defaulting.AgentLatestVersion)
 }
 
+func otelAgentImage() string {
+	// todo(mackjmr): make this dynamic once we have a non-dev image.
+	return "datadog/agent-dev:nightly-ot-beta-main"
```
[the remark this answers is not in the document]
Comment (mackjmr) [reply internal/controller/datadogagent/component/agent/default.go:109]: > Will this todo be resolved in this PR?
No, and defering to @dineshg13 for ETA of non-dev image.

File: cmd/main.go (hunk 4 above)
Comment (levan-m) [reply cmd/main.go:160]: Yes, staging shouldn't be using this, instead [annotation](https://github.com/DataDog/k8s-datadog-agent-ops/blob/main/charts/datadog-agent-with-operator/templates/datadogagent.yaml#L13-L15) on DDA.

File: internal/controller/datadogagent/component/agent/default.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (levan-m) [reply internal/controller/datadogagent/component/agent/default.go:109]: can this be moved with the rest of images constants
https://github.com/DataDog/datadog-operator/blob/756ed08d0d00ed329059aa5eccbfb4109c39f78b/pkg/defaulting/images.go#L25-L37

[the remark this answers is not in the document]
Comment (mackjmr) [reply internal/controller/datadogagent/component/agent/default.go:109]: Still discussing with Dinesh what image to use, I'll keep this in mind when I update it.

[the remark this answers is not in the document]
Comment (mackjmr) [reply internal/controller/datadogagent/component/agent/default.go]: @levan-m moved to `images.go` in https://github.com/DataDog/datadog-operator/pull/1559/commits/d1345c2ca9a0c2983a0bc696b402478ea3edf2d6

**Review by levan-m — APPROVED** [review]
Reviewed latest commits and looks good to me. Please verify the test cluster state isn't caused by this change before merging.

**Comment by wspurgin** [comment]
How would one enable OTEL prior to this PR if using the operator? I see that there's a command line flag, but it's unclear from the docs how to use those in a datadog manifest.

Relatedly, is there any target release planned that will include this feature?

**Comment by levan-m** [comment]
Hi @wspurgin, command line flag was added to enable internal testing using overrides section in the DatadogAgent CRD. This PR delivers feature support and will be released with v1.12, ETA early February.

**Comment by wspurgin** [comment]
Thanks for the response @levan-m - so am I right in my understanding that there _is not_ any way a user of the DD operator helm chart could enable the "beta" OTeL flag themselves?

Is there any concern with _allowing_ us to opt in for the time being (for users of the operator from versions v1.8 -> v1.11.1)? Otherwise I'll just have to drop the operator entirely and configure the agents in  Helm directly (the [operator is the recommended way](https://docs.datadoghq.com/containers/kubernetes/installation?tab=datadogoperator#installation), so I'd prefer obviously _not_ to undo my last week of work to do the non-recommended approach 😓)

**Comment by levan-m** [comment]
@wspurgin, there are ways to enabled it (render helm chart and edit manifest or use annotation added here https://github.com/DataDog/datadog-operator/pull/1475)  however main challenge will be correctly configuring the OTEL agent.

I'd recommend waiting till next week for 1.12 release candidate and use the feature added with this PR. For the RC setup you will need to install CRDs manually as helm chart isn't getting updated until final release.

**Comment by wspurgin** [comment]
Thanks @levan-m - I'll just wait for that RC. I tried the annotation (but the agent would crash with an error saying `otel-agent` didn't exist) - likely you said I'm probably missing some config for that to work properly. I'll be on the lookout for that RC once it's out! Thanks again 👍

**Pull request merged**